nestkernel/stopwatch.h

```diff
@@ -0,0 +1,505 @@
+     << ", diff: " << ( _end - _beg ) << ", prev: " << _prev_elapsed << ")";
+#endif
+  os << std::endl;
+#endif
+}
+
+template < clockid_t clock_type >
+inline size_t
+StopwatchBase< clock_type >::get_current_time()
+{
+  // We use a monotonic timer to make sure the stopwatch is not influenced by time jumps (e.g. summer/winter time).
+  timespec now;
+  clock_gettime( clock_type, &now );
+  return now.tv_nsec + now.tv_sec * timeunit_t::SECONDS;
+}
+
+template < clockid_t clock_type >
+inline std::ostream&
+operator<<( std::ostream& os, const StopwatchBase< clock_type >& stopwatch )
+{
+  stopwatch.print( "", timeunit_t::SECONDS, os );
+  return os;
+}
+
+}
+
+enum StopwatchVerbosity
+{
+  Normal,  //<! Always measure stopwatch
+  Detailed //<! Only measure if detailed stopwatches are activated
+};
+
+enum StopwatchType
+{
+  MasterOnly, //<! Only the master thread owns a stopwatch
+  Threaded    //<! Every thread measures an individual stopwatch
+};
+
+
+/** This is the base template for all Stopwatch specializations.
+ */
+/** Base timer class, which only measures a single timer, owned by the master thread. Might only actually measure time
+ * if detailed timers are enabled.
+ */
+template < StopwatchVerbosity detailed_timer, StopwatchType threaded_timer, typename = void >
+class Stopwatch
+{
+public:
+  void
+  start()
+  {
+#pragma omp master
+    {
+      walltime_timer_.start();
+      cputime_timer_.start();
+    }
+  }
+
+  void
+  stop()
+  {
+#pragma omp master
+    {
+      walltime_timer_.stop();
+      cputime_timer_.stop();
+    }
+  }
+
+  bool
+  isRunning() const
+  {
+    bool isRunning = false;
+#pragma omp master
+    {
+      isRunning = walltime_timer_.isRunning();
+    };
+    return isRunning;
+  }
+
+  double
+  elapsed( timers::timeunit_t timeunit = timers::timeunit_t::SECONDS ) const
+  {
+    double elapsed = 0.;
+#pragma omp master
+    {
+      elapsed = walltime_timer_.elapsed( timeunit );
+    };
+    return elapsed;
+  }
+
+  void
+  reset()
+  {
+#pragma omp master
+    {
+      walltime_timer_.reset();
+      cputime_timer_.reset();
+    }
+  }
+
+  void
+  print( const char* msg = "",
+    timers::timeunit_t timeunit = timers::timeunit_t::SECONDS,
+    std::ostream& os = std::cout ) const
+  {
+#pragma omp master
+    walltime_timer_.print( msg, timeunit, os );
+  }
+
+  void
+  output_timer( DictionaryDatum& d, const Name& walltime_name, const Name& cputime_name )
+  {
+    def< double >( d, walltime_name, walltime_timer_.elapsed() );
+    def< double >( d, cputime_name, cputime_timer_.elapsed() );
+  }
+
+private:
+  timers::StopwatchBase< CLOCK_MONOTONIC > walltime_timer_;
+  timers::StopwatchBase< CLOCK_THREAD_CPUTIME_ID > cputime_timer_;
+};
+
+/** If the user deactivated detailed timers, Stopwatch instance with the detailed flag will become an empty Stopwatch,
+ * which will be safely ignored by the compiler, as if the instance was not declared (e.g. as a member).
```

Comment:
Doxygen comments should start with a brief one-liner, then an empty line, then the full info.